### PR TITLE
Added several matrix tests, many of them XFail'ing

### PR DESCRIPTION
**Pull request by MrTrillian**
This is the result of some of my investigations about our support for matrices, especially around matrix orientations. Some of the tests pass, others demonstrate issues #1780, #1781, #1788, #1790 and #1795 and are marked with XFail.

### CODE REVIEW

File: tools/clang/test/CodeGenHLSL/quick-test/matrix_orientation_input.hlsl [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// RUN: %dxc /T vs_6_0 /E main %s | FileCheck %s
+
+// Test reading input matrices in both orientations.
+
+int main(row_major int4x4 r : R, column_major int4x4 c : C) : OUT
+{
+    // Unfortunately the loads get reordered here.
```
Comment (MrTrillian) [inline tools/clang/test/CodeGenHLSL/quick-test/matrix_orientation_input.hlsl]: Should I split this into two tests? I don't like to depend on a seemingly random reordering.

File: tools/clang/test/CodeGenHLSL/quick-test/matrix_orientation_transpose_in_function.hlsl [hunk 2]
```diff
@@ -0,0 +1,16 @@
+// RUN: %dxc /T vs_6_0 /E main > %s | FileCheck %s | XFail
+
+// Regression test for a bug where the transpose isn't performed,
```
Comment (vcsharma) [inline tools/clang/test/CodeGenHLSL/quick-test/matrix_orientation_transpose_in_function.hlsl]: better to mention bug id for easy reference later on.

**Comment by vcsharma** [comment]
This may or may not be directly related to orientation, but here is another one:

```
float4 vs_main() : SV_Position
{
    row_major float2x2 a = {1, 2, 3, 4};
    float r[4];
    r = (float[4]) a;
    return (float4) r;
}
```

dxc output:

```
call void @dx.op.storeOutput.f32(i32 5, i32 0, i32 0, i8 0, float 1.000000e+00)  ; StoreOutput(outputSigId,rowIndex,colIndex,value)
  call void @dx.op.storeOutput.f32(i32 5, i32 0, i32 0, i8 1, float 1.000000e+00)  ; StoreOutput(outputSigId,rowIndex,colIndex,value)
  call void @dx.op.storeOutput.f32(i32 5, i32 0, i32 0, i8 2, float 1.000000e+00)  ; StoreOutput(outputSigId,rowIndex,colIndex,value)
  call void @dx.op.storeOutput.f32(i32 5, i32 0, i32 0, i8 3, float 1.000000e+00)  ; StoreOutput(outputSigId,rowIndex,colIndex,value)
```

fxc output:

`mov o0.xyzw, l(1.000000,2.000000,3.000000,4.000000)`

File: tools/clang/test/CodeGenHLSL/quick-test/matrix_orientation_transpose_in_function.hlsl (hunk 2 above)
Comment (MrTrillian) [reply tools/clang/test/CodeGenHLSL/quick-test/matrix_orientation_transpose_in_function.hlsl]: Addressed in latest commit.

**Pull request merged**